pkg/cmd/checkendpoints/controller/connection_checker.go

```diff
@@ -0,0 +1,244 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	operatorcontrolplanev1alpha1 "github.com/openshift/api/operatorcontrolplane/v1alpha1"
+	"github.com/openshift/library-go/pkg/operator/events"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/cmd/checkendpoints/operatorcontrolplane/podnetworkconnectivitycheck/v1alpha1helpers"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/cmd/checkendpoints/trace"
+)
+
+// ConnectionChecker checks a single connection and updates status when appropriate
+type ConnectionChecker interface {
+	Run(ctx context.Context)
+	Stop()
+}
+
+// NewConnectionChecker returns a ConnectionChecker.
+func NewConnectionChecker(check *operatorcontrolplanev1alpha1.PodNetworkConnectivityCheck, client v1alpha1helpers.PodNetworkConnectivityCheckClient, recorder events.Recorder) ConnectionChecker {
+	return &connectionChecker{
+		check:    check,
+		client:   client,
+		recorder: recorder,
+		stop:     make(chan interface{}),
+	}
+}
+
+type connectionChecker struct {
+	check       *operatorcontrolplanev1alpha1.PodNetworkConnectivityCheck
+	client      v1alpha1helpers.PodNetworkConnectivityCheckClient
+	recorder    events.Recorder
+	updatesLock sync.Mutex
+	updates     []v1alpha1helpers.UpdateStatusFunc
+	stop        chan interface{}
+}
+
+// add queues status updates in a queue.
+func (c *connectionChecker) add(updates ...v1alpha1helpers.UpdateStatusFunc) {
+	c.updatesLock.Lock()
+	defer c.updatesLock.Unlock()
+	c.updates = append(c.updates, updates...)
+}
+
+// checkConnection checks the connection every second, updating status as needed
+func (c *connectionChecker) checkConnection(ctx context.Context) {
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+	defer klog.V(1).Infof("Stopped connectivity check %s.", c.check.Name)
+	for {
+		select {
+		case <-c.stop:
+			return
+		case <-ctx.Done():
+			return
+
+		case <-ticker.C:
+			go func() {
+				c.checkEndpoint(ctx, c.check)
+				c.updateStatus(ctx)
+			}()
+		}
+	}
+}
+
+// Run starts the connection checker.
+func (c *connectionChecker) Run(ctx context.Context) {
+	ctx2, cancel := context.WithCancel(ctx)
+	go func() {
+		select {
+		case <-c.stop:
+			cancel()
+		case <-ctx2.Done():
+		}
+	}()
+	go wait.UntilWithContext(ctx2, func(ctx context.Context) {
+		c.checkConnection(ctx2)
+	}, 1*time.Second)
+	klog.V(1).Infof("Started connectivity check %s.", c.check.Name)
+}
```

Comment:
run methods normally block on context

Reply:
Done. (https://github.com/openshift/cluster-kube-apiserver-operator/pull/846#discussion_r439068438 is the other side of this)